internal/testrunner/runners/pipeline/test_result.go

```diff
@@ -222,6 +223,35 @@ func unmarshalTestResult(body []byte) (*testResult, error) {
 	return &tr, nil
 }
 
+// jsonUnmarshalUsingNumber is a drop-in replacement for json.Unmarshal that
+// does not default to unmarshaling numeric values to float64 in order to
+// prevent low bit truncation of values greater than 1<<53.
+// See https://golang.org/cl/6202068 for details.
+func jsonUnmarshalUsingNumber(data []byte, v interface{}) error {
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.UseNumber()
+	err := dec.Decode(v)
+	if err != nil {
+		if err == io.EOF {
+			return errors.New("unexpected end of JSON input")
+		}
+		return err
+	}
+	// Make sure there is no invalid syntax after the message
+	// to match json.Unmarshal's behaviour.
+	remains, err := io.ReadAll(dec.Buffered())
+	if err != nil {
+		return err
+	}
+	for _, b := range remains {
+		if b > ' ' || (b != ' ' && b != '\t' && b != '\r' && b != '\n') {
+			// Mimic encoding/json error for this case, but without rigmarole.
+			return fmt.Errorf("invalid character %q after top-level value", b)
```

Comment:
No, this won't work. Some trailing bytes are acceptable, others are not; a JSON stream can have arbitrarily large quantities of whitespace outside literals and this extends to trailing and leading data.

Reply:
`dec.More()` ignores whitespaces, are there other acceptable bytes? The code I propose passes the test cases added in this PR (after adapting the error strings).

Reply:
Or something like this could be done to keep this error message:
```
        _, err = dec.Token()
        if err != nil && !errors.Is(err, io.EOF) {
                r, ok := dec.Buffered().(io.ByteReader)
                if !ok {
                        return err
                }
                b, err := r.ReadByte()
                if err != nil {
                        return err
                }
                // Mimic encoding/json error for this case, but without rigmarole.
                return fmt.Errorf("invalid character %q after top-level value", b)
        }
```

Reply:
>`dec.More()` ignores whitespaces

Nice. I didn't notice that. To retain the error parity we'll need something like the second option or what I have. I'm not convinced that the assertion to `io.ByteReader` is simpler of lighter though. We expect that in the happy case that remaining will be ~0, and in the unhappy case remaining will still be ~small (it must have fit in memory), but we cannot guarantee that `dec.Buffered()` will return an `io.ByteReader`, so we may miss cases.

Reply:
What I would like to avoid is to do parsing here (the `if b > ' ' || (b != ' ' && b != '\t' && b != '\r' && b != '\n')` line), and rely on the decoder for this.

Regarding the error message, I don't have a strong opinion, but at this level I think that it is ok if we don't show the found character.

Reply:
OK. This will complicate the test code. I will do it next week.

Reply:
No need to complicate test code, actually I think that we could simplify it, and only check that it returns an error when expected, without checking the error message.